File: 0894-all-possible-full-binary-trees/0894-all-possible-full-binary-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var allPossibleFBT = function(n) {
  return !(n % 2) ? [] : helper(n);
};

const helper = (n) => {
  if (n === 1) return [new TreeNode(0)];
  
  const ans = [];
  for (let i = 1; i < n; i += 2) {
    const left = helper(i);
    const right = helper(n - 1 - i);
    for (let l of left) {
      for (let r of right) {
        const node = new TreeNode(0, l, r);
        ans.push(node);
      }
    }
  }
  
  return ans;
}

File: 0894-all-possible-full-binary-trees/0894-all-possible-full-binary-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        @lru_cache(None)
        def helper(n):
            if n == 1:
                return [TreeNode(0)]
              
            ans = []
            for i in range(1, n, 2):
                left = helper(i)
                right = helper(n - 1 - i)
                for l in left:
                    for r in right:
                        node = TreeNode(0, l, r)
                        ans.append(node)
            return ans
          
          
        return [] if not n % 2 else helper(n)